pubsub/azurepubsub/example_test.go

```diff
@@ -117,3 +121,80 @@ func ExampleOpenSubscription() {
 	// See https://godoc.org/github.com/Azure/azure-service-bus-go#Message.Complete.
 	msg.Ack()
 }
+
+func ExampleSubscriptionWithAutoDeleteAndNoAck() {
+
+	ctx := context.Background()
+	// See docs below on how to provision an Azure Service Bus Namespace and obtaining the connection string.
+	// https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-dotnet-get-started-with-queues
+	connString := os.Getenv("SERVICEBUS_CONNECTION_STRING")
+	topicName := "test-topic"
+	subscriptionName := "test-sub"
+
+	if connString == "" {
+		log.Fatal("Service Bus ConnectionString is not set")
+	}
+
+	// Construct a Service Bus Namespace from a SAS Token.
+	// See https://godoc.org/github.com/Azure/azure-service-bus-go#Namespace.
+	ns, err := azurepubsub.NewNamespaceFromConnectionString(connString)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Construct a Service Bus Topic for a topicName associated with a NameSpace.
+	// See https://godoc.org/github.com/Azure/azure-service-bus-go#Topic.
+	sbTopic, err := azurepubsub.NewTopic(ns, topicName, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer sbTopic.Close(ctx)
+
+	// Construct Receiver to AutoDelete messages.
+	// See https://godoc.org/github.com/Azure/azure-service-bus-go#SubscriptionWithReceiveAndDelete.
+	var opts []servicebus.SubscriptionOption
+	opts = append(opts, servicebus.SubscriptionWithReceiveAndDelete())
+
+	// Construct a Service Bus Subscription which is a child to a Service Bus Topic.
+	// See https://godoc.org/github.com/Azure/azure-service-bus-go#Topic.NewSubscription.
+	sbSub, err := azurepubsub.NewSubscription(sbTopic, subscriptionName, opts)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer sbSub.Close(ctx)
+
+	// Construct a custom Ack Handler to override the driver acking.
+	// This is useful for a number of reasons including noAcks or custom acks.
```

Comment:
Isn't that overriding? I use the term override because it clearly conveys the result behavior - there is no second guessing as to its intention. Also, I recently added bulk disposition to the Service Bus SDK, app devs may want to use this to throttle acking or custom log & ack - this approach solves both of these scenarios

Reply:
It's not overriding:
- If you provide an ack func, that means you're in receive-and-delete mode, so there is no system ack function (question: what happens when you call Complete in this case)?
- The only way to get the "real" ack function is to _not_ provide your own. 

We just want to solve the problem of what to do with an at-most-once system. We don't want to provide general hooks for customizing behavior at this time. We'd rather make the client smart enough so that that isn't necessary. We want to minimize the number of options and hooks.

Reply:
Got it - calling Complete (or any DispositionAction) in at-most-once mode is disregarded by Service Bus. Once its compete (read once), its complete - you cant revoke the state.
 
Because of this, there is no harm calling driver.Ack, just wasted round trip. I understand your approach and will refactor to solve the immediate issue.